### PR TITLE
Estimating Memory Overhead Before Domain Creation

### PR DESCRIPTION
**Pull request by OhmSpectator**
This PR introduces an advanced approach to memory management during the creation of new domains within the ZedManager component. The primary innovation lies in the pre-estimation of memory overhead required by a new domain, ensuring the device has sufficient memory capacity to support both the domain's requirements and associated overhead.

## Key Changes
- **Memory Overhead Calculation**: Implemented in `updatestatus.go` utilizing the `CountMemOverhead` function from the hypervisor package.
- **Memory Usage Accounting**: Updated `getRemainingMemory` in `memorysizemgmt.go` to include memory overhead in the computation of used memory size.
- **Hypervisor Package Extensions**: Adapted to support memory overhead estimation across different hypervisor types.
- **Assignable Adapters Integration**: Added adapter information to `zedmanagerContext`, including a new subscription within `zedmanager.go`. This integration is critical for precise memory overhead estimations.

## Proposed Steps for Moving Forward
- **Step One**: Maintain the current PR, which incorporates `memOverhead` into the code. This update aims to mitigate the issue of initiating multiple large virtual machines, a challenge we currently face.
- **Step Two**: Backport this enhancement to the LTS branches.
- **Step Three**: Address the more complex issue of latent memory usage in a separate PR. This topic requires further discussion as it presents a nuanced challenge and does not offer an immediate resolution to our existing issues.

### CODE REVIEW

**Comment by OhmSpectator** [comment]
I have not tested it yet, just made the code buildable

**Comment by eriknordmark** [comment]
I have an overall behaviorial question: does the accounting of memory in getRemainingMemory() for an app instance change as the app instance is running, or is it fully defined by the assigned adapters, type of hypervisors, and the amount of memory allocated to the VM/container?
Reason I'm asking is that we want to make sure that if a user can deploy three app instances in order A, B, C without getRemainingMemory showing that there isn't enough for C, this should be stable independent of order that the those app instances are deployed and/or booted.

File: pkg/pillar/cmd/zedmanager/zedmanager.go [hunk 1]
```diff
@@ -539,6 +567,38 @@ func checkDelayedStartApps(ctx *zedmanagerContext) {
 	}
 }
 
+func handleAA(ctx *zedmanagerContext, status types.AssignableAdapters, key string) {
+	log.Functionf("handleAA(%s)", status.Key())
+	if key != "global" {
+		log.Functionf("handleAAImpl: ignoring %s", key)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedmanager/zedmanager.go:573]: Some instances of "Impl" are still in log messages.

**Comment by OhmSpectator** [comment]
> does the accounting of memory in getRemainingMemory() for an app instance change as the app instance is running, or is it fully defined by the assigned adapters, type of hypervisors, and the amount of memory allocated to the VM/container?

My current understanding is that the MemOverhead values calculated are static. The MemOverhead function estimates the potential memory overhead for a VM based on the configuration and list of assignable adapters. This estimation is done once for each app instance and does not change dynamically during the app's runtime. @rucoder can provide some more explanation here.

While the MemOverhead itself is static, the way getRemainingMemory() calculates available memory does change with this update. getRemainingMemory() is dynamic, assessing the system's memory availability in real-time. The PR's changes affect this function by adding the static MemOverhead value to the total memory requirement calculation for an app instance. This means that while the overhead value is a constant once calculated, its inclusion is a new factor in the dynamic calculation performed by getRemainingMemory(). However, the "variation" part is not changed within this PR.

In summary, the PR introduces a static memory overhead estimation into the dynamic memory availability calculation performed by getRemainingMemory()

File: pkg/pillar/cmd/zedmanager/memorysizemgmt.go [hunk 2]
```diff
@@ -29,6 +29,7 @@ func getRemainingMemory(ctxPtr *zedmanagerContext) (uint64, uint64, uint64, erro
 	for _, st := range itemsAppInstanceStatus {
 		status := st.(types.AppInstanceStatus)
 		mem := uint64(status.FixedResources.Memory) << 10
+		mem += status.MemOverhead
```
Comment (rouming) [inline pkg/pillar/cmd/zedmanager/memorysizemgmt.go:32]: I would go further with the proposed change. The 'haltingMemorySize' is incorrect and dangerously racy: an app is marked as "Active = false", but qemu process is still alive (takes time to bring it down), but we already start allocating memory for a new app, considering that memory should be enough. That's bad. I propose avoid the 'halting' calculations at all, and always account "vRAM + overhead" memory for all existent apps.

File: pkg/pillar/cmd/zedmanager/updatestatus.go [hunk 3]
```diff
@@ -577,6 +578,30 @@ func doActivate(ctx *zedmanagerContext, uuidStr string,
...
+		// If we have not yet calculated memory overhead - do it now
+		if status.MemOverhead == 0 {
+			// Get hypervisor
+			hyp, err := hypervisor.GetHypervisor(*ctx.hypervisorPtr)
+			if err != nil {
+				log.Fatalf("Cannot get hypervisor: %s", err)
+			}
+
+			// Create a mock domain config to calculate memory overhead
+			mockDomainConfig := types.DomainConfig{
```
Comment (rouming) [inline pkg/pillar/cmd/zedmanager/updatestatus.go:590]: lookupDomainConfig() instead ?

File: pkg/pillar/cmd/zedmanager/zedmanager.go [hunk 4]
```diff
@@ -375,6 +384,22 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject, ar
 	ctx.subVolumesSnapStatus = subVolumesSnapshotStatus
 	_ = subVolumesSnapshotStatus.Activate()
 
+	ctx.subAssignableAdapters, err = ps.NewSubscription(pubsub.SubscriptionOptions{
```
Comment (rouming) [inline pkg/pillar/cmd/zedmanager/zedmanager.go:387]: This probably is not needed if lookupDomainConfig() is used.

File: pkg/pillar/cmd/zedmanager/memorysizemgmt.go (hunk 2 above)
Comment (rouming) [reply pkg/pillar/cmd/zedmanager/memorysizemgmt.go:32]: Also the `sysTotalMemory()` (used for `deviceMemorySize`) seems not very reliable: asynchronous pub/sub with 25 seconds period (see `metricsTimerStart()`). In 25 seconds a couple of fat Vms can be started, but calculated `deviceMemorySize` won't account this, which is again bad. Sys memory calculations should very simple and reliable: request true sys  memory counters just from the code accessing the /proc synchronously, that will guarantee that we see current state of the system

**Comment by rouming** [comment]
This should be a very good work, which fixes incorrect free app memory estimation. @OhmSpectator I added a few comments requesting go deeper with the fix (not just memory overhead calculations). Feel free to address these request in the following PRs.

File: pkg/pillar/cmd/zedmanager/updatestatus.go (hunk 3 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/updatestatus.go:590]: The problem is that it does not yet exist for the VM that we are creating. We create the config later in this function on MayebeAddDomainConfig (or something similar) only after we are sure we have enough recourses.

File: pkg/pillar/cmd/zedmanager/zedmanager.go (hunk 4 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/zedmanager.go:387]: 1. As I explained before, the moment when we evaluate the memory requirements is earlier than the moment when DomainConfig is created.
2. We still need to have a list of assignable adapters that are stored per system, not per domain. As far as I understand we need this system-wide list to check against it the list of the adapters assigned in the domainConfig. @rucoder should know this logic. I'll recheck how it's used. So we have to get this list in zedmanager.

File: pkg/pillar/cmd/zedmanager/memorysizemgmt.go (hunk 2 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/memorysizemgmt.go:32]: > I propose avoid the 'halting' calculations at all, and always account "vRAM + overhead" memory for all existent apps.

For me, it makes sense. I'll recheck how this halting memory is used and why it's separated from the "occupied" memory. Maybe @eriknordmark can give some clarifications here.

> Sys memory calculations should very simple and reliable: request true sys memory counters just from the code accessing the /proc synchronously, that will guarantee that we see current state of the system

I have not checked the implementation of the sysTotalMemory() yet. Will do. But your suggestion looks meaningful to me.

File: pkg/pillar/cmd/zedmanager/updatestatus.go (hunk 3 above)
Comment (rouming) [reply pkg/pillar/cmd/zedmanager/updatestatus.go:590]: You should be able to have `AppInstanceConfig` in you hands received from the `lookupAppInstanceConfig`. (the config should be completely parsed to this point): so `AppInstanceConfig.IoAdapterList` and other fixed resourcs. So probably to make the function cleaner (not mocking fake domain config) makes sense to adapt the `hypervisor.CountMemOverhead()` to the set of parameters and not a complete domain config. Just thoughts.

Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/updatestatus.go:590]: Yeah, sure. I can also introduce a new structure for passing this args. I just didn't want to change the signatures of the original functions used to estimate the overhead (they are using exactly DomainConfig). But maybe it is worth it.

File: pkg/pillar/cmd/zedmanager/memorysizemgmt.go (hunk 2 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedmanager/memorysizemgmt.go:32]: @OhmSpectator I didn't even know there was a separate halting size. What does git blame say abouts its history?

We do want to trigger a retry  (of apps which couldn't started due to lack of memory) when an app has halted, but that doesn't require an exact number, and as Roman said, the retry can't be done until it is halted.

**Comment by OhmSpectator** [comment]
In the CountMemOverhead, use explicit params instead of DomainConfig

File: pkg/pillar/cmd/zedmanager/updatestatus.go (hunk 3 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/updatestatus.go]: I've switched to explicit parameters instead of DomainConfig.

File: pkg/pillar/cmd/zedmanager/memorysizemgmt.go (hunk 2 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/memorysizemgmt.go:32]: @eriknordmark actually, it was added by you =D
https://github.com/lf-edge/eve/commit/51ffeb1316727a3ad1b444f39adff29a068f6ebb
I'll try to understand the concept now.

Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/memorysizemgmt.go:32]: Here is my current understanding of the different "memories" used when when we decide wether ti run a new domain:

### usedMemorySize

**Description**: This represents the total memory used by applications that are either activated or in the process of being activated.

**Details**: It's a cumulative measure of the memory requirements of all running or activating applications. This includes both the RAM allocated to each application (`FixedResources.Memory`) and any additional overhead (`MemOverhead`).

### haltingMemorySize

**Description**: A subset of `usedMemorySize`, representing the memory used by applications that are currently in the process of halting.

**Details**: This memory is allocated to applications that have been initiated for shutdown but haven't completed the process yet. Although part of `usedMemorySize`, it's distinguished because this memory is expected to be freed up shortly as the applications complete their halting process.

### latentMemorySize

**Description**: Memory used by applications that are halted, meaning they are neither running nor in the process of activating.

**Details**: This category covers applications that are in a passive state, not actively using resources but still reserving memory. This memory isn't immediately available but may become so if these applications are completely shut down or remain inactive.

### memoryReservedForEve

**Description**: Memory reserved specifically for the EVE system itself.

**Details**: This includes the memory limit set globally for EVE (`EveMemoryLimitInBytes`). If the system uses ZFS for persistence (`PersistZFS`), it also includes the maximum size of the ZFS ARC (`zfsArcMaxLimit`), which is an additional memory reservation for file system caching.

### deviceMemorySize

**Description**: The total memory available on the device.

**Details**: This represents the physical memory capacity of the device. It corresponds to the `TotalMemoryMB` metric sent from the DomainManager and is equivalent to the `MemTotal` record found in `/proc/meminfo` on the device.

### remainingMemory

**Description**: The amount of memory available for new applications or processes.

**Details**: Calculated as the total device memory (`deviceMemorySize`) minus the total used memory (sum of `usedMemorySize` and `memoryReservedForEve`). It represents the current available memory capacity that can be allocated to new or additional applications.

Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/memorysizemgmt.go:32]: And here is how we decide whether to activate a domain or not.

### 1. Calculate Memory Requirements
- **Determine Total Need (`need`)**: Calculate the total memory requirement for the new domain, incorporating fixed memory allocation (`FixedResources.Memory`) and additional overhead (`MemOverhead`).

### 2. Determine Available Memory
- **Calculate `remainingMemory`**: This is the total device memory (`deviceMemorySize`) minus the used memory (`usedMemorySize`) and memory reserved for EVE (`memoryReservedForEve`). Note that halting memory is not considered free here. But latent - is. 

### 3. Decision Making
- **Sufficient Memory Without Halting Memory**: 
  - If `remainingMemory` ≥ `need`, there's enough memory to start the new domain.
  - Warning for potential overcommitment is logged if `remainingMemory` < `need` + `latentMemorySize`. **I find this current behaviour dangerous as it risks system stability by potentially overloading memory resources.** 

- **Potential Sufficiency with Halting Memory**: 
  - If `remainingMemory` < `need`, but `remainingMemory` + `haltingMemorySize` ≥ `need`, delay the start of the new domain.
  - Log a notice, set error severity to notice, and specify `retryCondition` for delayed start.

- **Insufficient Memory Even with Halting Memory (With Running Apps)**: 
  - This scenario occurs when `remainingMemory` + `haltingMemorySize` < `need`, but other applications are running.
  - The presence of running applications suggests potential future availability of memory, leading to a less critical view of the current memory shortage.
  - Set error severity to warning, indicating a non-critical issue that might resolve with the completion of running applications.
  - Delay the start of the new domain and specify `retryCondition` indicating a retry when memory becomes available.

- **Insufficient Memory Even with Halting Memory (No Running Apps)**: 
  - If `remainingMemory` + `haltingMemorySize` < `need` and no applications are running, set the error severity to error.
  - Delay the start of the new domain and specify `retryCondition` for when more memory is available.

Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/memorysizemgmt.go:32]: @eriknordmark, @rouming, my current proposal is:

Step one. We leave the current PR as it is, which adds memOverhead to the code and can potentially prevent the issue of starting multiple large virtual machines that we are currently experiencing.

Step two. We address the underlying issue of latent memory usage in a separate PR. This may require further discussion as it is a more complex problem and won't immediately solve any of the existing issues we are facing.

File: pkg/pillar/cmd/zedmanager/zedmanager.go (hunk 1 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/zedmanager.go]: Fixed

**Comment by OhmSpectator** [comment]
Fixed typos in log messages and the Yetus errors

File: pkg/pillar/cmd/zedmanager/memorysizemgmt.go (hunk 2 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedmanager/memorysizemgmt.go:32]: > is equivalent to the `MemTotal` record found in `/proc/meminfo` on the device.

If you plan to use your explanation as documentation, then the above needs more text. If Kvm is is correct, but if Xen then /proc/meminfo is only what is available in dom0 and EVE uses xen tools specific things to find the total memory of the device.

Comment (eriknordmark) [reply pkg/pillar/cmd/zedmanager/memorysizemgmt.go:32]: > @eriknordmark, actually, it was added by you here [51ffeb1](https://github.com/lf-edge/eve/commit/51ffeb1316727a3ad1b444f39adff29a068f6ebb) =D I'll try to understand the concept now.

Thanks for tracking that down.
Looks like haltingMemorySize is only used to product a slightly different error, hence it does not need to be that accurate. We will trigger a retry when memory when the app instance has halted no matter what.

File: pkg/pillar/types/zedmanagertypes.go [hunk 5]
```diff
@@ -263,7 +263,8 @@ type AppInstanceStatus struct {
 	// Effective time, when the application should start
 	StartTime time.Time
 	// Snapshot related information
-	SnapStatus SnapshottingStatus
+	SnapStatus  SnapshottingStatus
+	MemOverhead uint64 // Overhead for memory for VM
```
Comment (eriknordmark) [inline pkg/pillar/types/zedmanagertypes.go:267]: Please add the unit in the comment. (I guess it is MB.)

**Review by eriknordmark — APPROVED** [review]
LGTM.
It might make sense to add your explanation e.g., in pkg/pillar/docs/zedmanager.md or some other place.

Also, do we know whether these changes will cause us to fail to boot something which currently boots (but might result in an oom later.) It would be good to understand that so that we know whether some users might see an immediate failure with these changes where they today can run things but be at risk of oom.

**Comment by rene** [comment]
> LGTM. It might make sense to add your explanation e.g., in pkg/pillar/docs/zedmanager.md or some other place.
> 
> Also, do we know whether these changes will cause us to fail to boot something which currently boots (but might result in an oom later.) It would be good to understand that so that we know whether some users might see an immediate failure with these changes where they today can run things but be at risk of oom.

This PR looks very nice @OhmSpectator, thanks. I agree with @eriknordmark regarding add more documentation about it. Also, make sure to test it on ARM devices to avoid any surprises.... I can help with that, just let me know...

**Comment by OhmSpectator** [comment]
> It might make sense to add your explanation e.g., in pkg/pillar/docs/zedmanager.md or some other place.

@eriknordmark, do you mean the description of the memory types taken into consideration when making a decision on Domain activation, from https://github.com/lf-edge/eve/pull/3588#discussion_r1396008557 and https://github.com/lf-edge/eve/pull/3588#discussion_r1396086943 ? I think it would make sense to update documentation separately after we agree on the way we handle "latent" memory. 


> Also, do we know whether these changes will cause us to fail to boot something which currently boots (but might result in an oom later.)

Based on my understanding, this issue may only impact situations where users allocate nearly all of the device's memory to running applications. To my knowledge, there is only one such case, and this PR is a direct response to that situation.

File: pkg/pillar/cmd/zedmanager/memorysizemgmt.go (hunk 2 above)
Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/memorysizemgmt.go:32]: I put this explanation just to clarify the situation and to decide on the following steps with the PR: whether any extra actions were critical for addressing the original issue we were trying to resolve. Looks like they are not, but we still can make improvements. I'll create a ticket and put the comment there. Later I can rework it and make a part of the documentation.

File: pkg/pillar/types/zedmanagertypes.go (hunk 5 above)
Comment (OhmSpectator) [reply pkg/pillar/types/zedmanagertypes.go]: Fixed

**Comment by OhmSpectator** [comment]
I started to port the changes to 10.4 and realized that that version does not contain the changes done by @rucoder for better overhead estimation. Neither contains them 9.4.
So, I'm going to backport only the fix now.

File: pkg/pillar/cmd/zedmanager/memorysizemgmt.go (hunk 2 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedmanager/memorysizemgmt.go:32]: Great.

File: pkg/pillar/hypervisor/containerd.go [hunk 6]
```diff
@@ -49,6 +50,15 @@ func newContainerd() Hypervisor {
 	}
 }
 
+// CountMemOverhead - returns the memory overhead for a domain.
+// This implementation is used for Xen as well
+func (ctx ctrdContext) CountMemOverhead(domainName string, domainUUID uuid.UUID, domainRAMMemory int64, vmmMaxMem int64,
```
Comment (eriknordmark) [inline pkg/pillar/hypervisor/containerd.go:55]: I don't know how important this is, but as was reading the caller I saw that the order of in64 arguments is somewhat ad-hoc hence hard to remember:
1. memory
2. max memory
3. max cpus
4. cpus

It is sad that golang has no way to name arguments (unlike python) meaning that if you want named args you have to do a struct akin to
 foo(x, y, fooArgs{arg1: 27, arg2: 93})

Given the small number of callers of this function I don't propose that, but I like to complain about golang making things hard ;-)

**Review by eriknordmark — APPROVED** [review]
LGTM
